# 💻 프로미스
- 자바스크립트와 노드에서는 주로 비동기를 접한다.
- 특히 이벤트 리스너를 사용할 때 콜백 함수를 자주 사용한다.
- ES2015부터는 자바스크립트와 노드의 API들이 콜백 대신 프로미스(Promise)기반으로 재구성되며 콜백 지옥 현상을 극복했다.

```
state : pendfing -> fulfilled or reject
Producer vs Consumer
```

## 📌 Producer
#### `resolve`
>기능을 정상적으로 수행해서 마지막에 최종 데이터를 전달
#### `reject`
>기능을 수행하다가 중간에 문제가 생기면 호출

```
const promise = new Promise((resolve, reject) =>{
  console.log('doing something...');
  setTimeout(() => {
    1️⃣resolve('eunbi'); 
    2️⃣reject(new Error('no network'));
   
  },2000);
});
```

## 📌 Comsumers : then, catch , finally
1️⃣promise가 정상적으로 수행되어서 resolve 콜백함수에서 전달된 값(eunbi)가 들어옴 <br>
2️⃣실패시 'no network' 라는 에러발생<br>
#### `finally` 
>성공/ 실패 상관없이 항상 실행
```
promise.
then((value) => {
  console.log(value);
})
.catch(error=>{
  console.log(error);
})
.finally(()=> {console.log('finally');
});
```

## 📌 Promise chanining 
```
const fetchNumber = new Promise((resolve,reject)=>{
  setTimeout(()=> resolve(1),1000); // 프로미스 성공시 1초 후 1 전달
});

fetchNumber
.then(num => num*2) // 전달된 1*2 =2
.then(num => num*3) // 전달된 2*3 =6
.then(num => {
  return new Promise((resolve,reject)=>{ //새로운 프로미스
    setTimeout(()=> resolve(num - 1),1000); // 전달된 6-1 =5
  });
})
.then(num => console.log(num)); // 2초 후 5가 출력된다
```

## 📌 Error Handling
```
const getHen = () => 
  new Promise((resolve, reject)=>{
    setTimeout(() => resolve('🐓'), 1000);
  });

const getEgg = hen => 
  new Promise((resolve, reject)=>{
    setTimeout(() => reject(new Error(`error! ${hen} =>🥚`)), 1000);
  });

const cook = egg => 
  new Promise((resolve, reject)=>{
    setTimeout(() => resolve(`${egg} => 🍳`), 1000);
  });

getHen()
  .then(getEgg)
  .catch(error => {
    return '🥞'; 
  })
  .then(cook)
  .then(console.log)
  .catch(console.log);
```

### 🌊아이디와 비밀번호 입력시 name과 role 출력
```
class UserStorage{
  loginUser(id,password){
    return new Promise((resolve,reject) => {
      setTimeout(()=>{
        if(
          (id === 'eunbi' && password === '1234') ||
          (id === 'coder' && password === '2345')
        ) {
          resolve(id);
        }else{
          reject(new Error('not found'));
        }
      },2000);
      });
}

getRoles(user) {
  return new Promise((resolve,reject) => {
    setTimeout(()=>{
      if(user === 'eunbi'){
        resolve({ name : 'eunbi', role : 'admin'});
      }else{
        reject(new Error('no access'));
      }
     },1000);
   });
  }
}

const userStorage = new UserStorage();
const id = prompt('enter your id');
const password = prompt('enter your password');
userStorage.loginUser(id,password)
  .then(userStorage.getRoles)
  .then(user => alert(`Hello ${user.name}, you have a ${user.role} role`))
  .catch(console.log);
```

